### ISSUE TITLE
A gap between text input and keyboard on iOS

### ISSUE THREAD

**Issue by dmitryusikriseapps** [FaridSafi/react-native-gifted-chat#1306]
#### Issue Description

There is a gap between input text and keyboard on iOS. I tried to fix it with bottomOffset, but it didn't help.

#### Steps to Reproduce / Code Snippets

https://pastebin.com/pAJz5W2C
https://pastebin.com/RGCWP2JL
https://ibb.co/09NTN9Q

#### Expected Results

There shouldn't be nay gaps.

#### Additional Information

* Nodejs version: "8.12.0"
* React version: "16.8.3"
* React Native version: "0.59.9"
* react-native-gifted-chat version: "0.9.11"
* Platform(s) (iOS, Android, or both?): iOS
* TypeScript version: just pure JS

**Comment by nadav2051** [comment]
Also experiencing this:
![image](https://user-images.githubusercontent.com/6547468/61217944-8641f000-a719-11e9-81ba-3a033b02aee5.png)

**Comment by MaggieRezo** [comment]
I am having this same issue and am using SafeAreaView. Tested on my Xs max.

**Comment by aniciom** [comment]
I had a similar issue at #1294 i marked it as closed, but now with the new updates it happens again, even using that patch, any idea @xcarpentier ?

**Comment by Ariavm95** [comment]
It seems if you add "renderInputToolbar" prop, and only return "InputToolbar" component as the value of the prop, it will work fine.

**Comment by dmitryusikriseapps** [comment]
@Ariavm95 Hi, thank you for the advice, but it doesn't help

**Comment by hatemalimam** [comment]
any workaround here ? something is really wrong with the layout

**Comment by dmitryusikriseapps** [comment]
@hatemalimam https://pastebin.com/B8r0dHK2

**Comment by hatemalimam** [comment]
yes putting autoCorrect: false fixed the problem temporarily, but it's an annoying compromise though, sometimes also on some refresh, it gets broken again, not sure what's the behaviour here

**Comment by tconlin** [comment]
bottomOffset={50} worked for me. #1202

**Comment by llamaluvr** [comment]
I found this happens even on the latest version (0.11.2). It seems most likely to happen if I navigate to my chat screen, lock the phone, and then unlock it. It went away when I commented out the iPhone X logic to get around the home bar found at https://github.com/FaridSafi/react-native-gifted-chat/blob/e4576d4c4a2c92e52816488c58e54a0eaa628c4d/src/GiftedChat.tsx#L563. In my case, I already have iPhone X home bar avoidance logic in my parent views, so I think it was getting accounted for twice.

I'm of the opinion that home bar/ notch avoidance is so contextual that I prefer component libraries to not handle it and leave it to the implementing developer. In my case, I have a tab bar showing at the bottom sometimes that already needs to avoid the home bar.  I understand that having it in there may be more friendly for some, though, and may be more in line with the goals of the library. @xcarpentier I'd be happy to put in a pull request to fix, but want to do it in a way that's consistent with your outlook on this. I could see removing the iPhone X logic entirely, or leaving it on by default but providing a flag for users to turn it off.

**Comment by Miyou** [comment]
Just wanted to share, that this is what worked for me: `bottomOffset={isIphoneX() ? 125 : 55}`
Where isIphoneX() is:
```javascript
import { Dimensions, Platform } from 'react-native';
function isIphoneX() {
  const dimen = Dimensions.get("window");
  return (
    Platform.OS === "ios" &&
    !Platform.isPad &&
    !Platform.isTVOS &&
    (dimen.height === 812 ||
      dimen.width === 812 ||
      (dimen.height === 896 || dimen.width === 896))
  );
}
```
from npm package react-native-iphone-x-helper.

No clue why those numbers are the ones that work. The size of the gap seems linear with the distance from those numbers for me.

**Comment by ZaidAbo** [comment]
In my case the problem was the bottom tab navigator ,,, when i removed it ,, everything worked perfectly

**Comment by PaitoAnderson** [comment]
@ZaidAbo same here, I used the `tabBarVisible: false,` option in `react-navigation-tabs`

**Comment by skylus** [comment]
The same thing @nadav2051 describes still happens to me in 0.13.0 - ios. removed the tab bar. did anyone find a workaround ?

> Also experiencing this:
> ![image](https://user-images.githubusercontent.com/6547468/61217944-8641f000-a719-11e9-81ba-3a033b02aee5.png)
> 
> It looks to be happening when the keyboard changes it's side - when the auto correct component vanishes, the keyboard does not re-adjust which causes the GiftedChat component not to cover the entire screen leaving this gap
> 
> Edit: It seems to happen on iPhone XR and X, on iPhone 6, 8 it works perfectly, I suspect it might have something to do with the SafeAreaView

**Comment by abdielou** [comment]
I don't quite understand these lines of code and they might be the culprit for some people.

```
    if (isIphoneX()) {
      return bottomOffset === this._bottomOffset ? 33 : bottomOffset
    }
```
https://github.com/FaridSafi/react-native-gifted-chat/blob/master/src/GiftedChat.tsx#L589

Why is it being set to 33? The first render will leave my custom `bottomOffset` intact (83 in my case). Further renders will reset it to 33. This seems to cause the gap on iPhone X.

Also, I don't see any possible workaround as this is internally handled. Removing these lines fixes the issue for me.

Also, I'm not sending a PR for this as I'm still uncertain why this was added. thoughts?

**Comment by abdielou** [comment]
@JesuHrz 

Can you help explain the purpose of this change?

https://github.com/FaridSafi/react-native-gifted-chat/blob/master/src/GiftedChat.tsx#L589

**Comment by NMVikings** [comment]
You can hide your keyboard each time your app is going to background by using this code:

```
AppState.addEventListener('change', nextAppState => {
  if (nextAppState === 'background') {
    Keyboard.dismiss()
  }
})
```

But this is temporary hack! I think that GiftedChat should not ignore bottomOffset that developer passes to it. Or it should have to bottomOffsets: bottomOffset and bottomOffsetX, and if developer passes only bottomOffset, than summarise bottomOffset and 33, if there is bottomOffsetX to use only it.

**Comment by robert-narrate** [comment]
I am also having the same issue as @abdielou. the bottomOffset toggles between what I have set and 33

**Comment by alexpchin** [comment]
I currently have this to add to bottomOffset:

```js
export const CHAT_BOTTOM_OFFSET = () => {
  // When the keyboard opens, there seems to be a gap at the bottom of the input on iOS
  // We hide the bottomTabBar when the keyboard opens, the gap is the same height
  //
  // C.BOTTOM_TAB_BAR_HEIGHT -> iPhoneX 83 / iPhone8 59
  // C.NAVIGATION_BAR_HEIGHT ->  iPhoneX 88 / iPhone8 64
  // C.getStatusBarHeight(C.isIphoneX() -> iPhoneX 44 / iPhone8 20
  //
  // When coming back into the app when focused on the input on iPhoneX only there is a problem
  if (Platform.OS === 'ios') {
    if (isIphoneX()) {
      return getStatusBarHeight(isIphoneX()) + MIN_COMPOSER_HEIGHT; // iPhoneX = (44+33)=77
    } else {
      return HEADER_BAR_HEIGHT; // 44
    }
  }
  if (Platform.OS === 'android') return 0;
  return 0;
};
```

as well as:

```
// https://github.com/FaridSafi/react-native-gifted-chat/issues/1306#issuecomment-591956022
  const dismissKeyboardOnBackground = nextAppState => {
    if (nextAppState.match(/inactive|background/)) {
      // Close keyboard when the app goes into the background to fix an issue on iPhoneX where the bottomOffset is not being applied and whitespace added to the bottom.
      Keyboard.dismiss();
    }
  };

  useEffect(() => {
    AppState.addEventListener('change', dismissKeyboardOnBackground);
    return () => {
      AppState.removeEventListener('change', dismissKeyboardOnBackground);
    };
  }, []);
```

It seems to have resolved the problem but it's hurting my soul.

**Comment by JesuHrz** [comment]
I added the value of 33 because the default value of SafeAreaView is 34 and I subtracted 1 to make it fit better as you can see in the following pictures:

<p align="center">
  <img width="330" src="https://user-images.githubusercontent.com/28031187/77798152-5422b500-7040-11ea-8078-3bbd3493f6ef.png">
</p>
<p align="center">
  <img width="330" src="https://user-images.githubusercontent.com/28031187/77798144-52f18800-7040-11ea-843d-265d020d4e2e.png">
</p>

I did this in order to support devices with Notch.

I see some of you are adding the GiftedChat component as a BottomTabNavigator component and in which you are not contemplating the height of the size of the BottomTabNavigator, which by default in react-navigation is 50
DEFAULT_TABBAR_HEIGHT = 50

Then not to contemplate it they must add

bottomOffset = DEFAULT_TABBAR_HEIGHT + SafeAreaView Hight

The size of SafeAreaView can be taken out in this way 
```
import { getInset } from 'react-native-safe-area-view';
```

and this library is already installed with react-navigation so you can access it without having to reinstall it

**Comment by yasirtrimulabs** [comment]
> react-native-safe-area-view

thanks but getInset is gone from the 'react-native-safe-area-view' and this library was not installed with react-navigation 5 
Alternatively I found we can use 'react-native-safe-area-context' which is installed with react-navigation 5
`import {useSafeArea} from 'react-native-safe-area-context';
const insets = useSafeArea();
const SAFEAREA_HEIGHT = insets.bottom;
`

**Comment by thrillhop** [comment]
This worked for me. I set a const for footerHeight that I pass to my navigation logic, and then use it as apart of the input offset: 

```
bottomOffset={
    isIphoneX() ?
    getBottomSpace() + Variables.footerHeight :
    Variables.footerHeight
}
```

**Comment by yasirtrimulabs** [comment]
> > react-native-safe-area-view
> 
> thanks but getInset is gone from the 'react-native-safe-area-view' and this library was not installed with react-navigation 5
> Alternatively I found we can use 'react-native-safe-area-context' which is installed with react-navigation 5
> 
> ```
> import {useSafeArea} from 'react-native-safe-area-context';
> const insets = useSafeArea();
> const SAFEAREA_HEIGHT = insets.bottom;
> ```

The bug still appears randomly.
So I than tried 
```
import { getBottomSpace } from 'react-native-iphone-x-helper'

getBottomSpace()
```
And in your giftedchat set bottomOffset like this
```
bottomOffset={getBottomSpace() + DEFAULT_TABBAR_HEIGHT}
```
And for react-navigation 5 after searching alot I used DEFAULT_TABBAR_HEIGHT = 49 instead of 50 now so far it's not appeared again.
🥂

**Comment by Hirbod** [comment]
Try the new version. I supplied a fix for that which should cover the issue.

**Comment by abdielou** [comment]
> Try the new version. I supplied a fix for that which should cover the issue.

@Hirbod Can you reference this issue in your fix?

**Comment by nickblumenthal** [comment]
In case others find this with a similar issue, the commit @abdielou was asking about is here:
https://github.com/FaridSafi/react-native-gifted-chat/commit/dcf8f0a23ad003f3773b653f630b01707a821ab9#

Unfortunately, I'm still having an occasional issue where multiple keyboardWillShow events are detected by GiftedChat even if the keyboard is already open.  This can happen if the user starts typing and then hits the input bar again.  This sets the internal `._bottomOffset` that is used to help calculate screen size each time an event is detected.  Unfortunately, if you don't set the `bottomOffset` prop this can lead to a state where the internal `._bottomOffset` is correctly set initially only to be set to zero by the second event.

![image](https://user-images.githubusercontent.com/9689775/120048243-6cd38580-bfcb-11eb-8425-3b6cf0e19713.png)

**Comment by nickblumenthal** [comment]
I'm solving this by setting `bottomOffset: getBottomSpace()` in the GiftedChat props.  This should manually override for notched IphoneX phones and let chat control it the rest of the time.

**Comment by tmhao2005** [comment]
For someone who uses `react-native-safe-area-context` from the version `^3.x`, the solution is to use hook like this:

```js
import { useSafeAreaInsets } from "react-native-safe-area-context";

function App() {
  const insets = useSafeAreaInsets();
  
  return (
    <GiftedChat
       // ...
       bottomOffset={insets.bottom}
    />
  )
}
```

**Comment by sumanbhattarai** [comment]
If you have a bottom tab navigator, then this might be a useful solution.

```
import { useBottomTabBarHeight } from "@react-navigation/bottom-tabs";

const App = () => {
const DEFAULT_TABBAR_HEIGHT = useBottomTabBarHeight();

return (
<GiftedChat
        // other props
        bottomOffset={DEFAULT_TABBAR_HEIGHT}
      />
     )
     }
```

**Comment by irodeanu** [comment]
`import { useSafeAreaInsets } from 'react-native-safe-area-context';`
`const insets = useSafeAreaInsets()`

and 

`
       <GiftedChat
        wrapInSafeArea={false}
        bottomOffset={insets.bottom}
... />
`

**Comment by 2DTW** [comment]
@irodeanu Thank you. This is the answer that works best for me. Also eliminated a random empty gap at the top of the gifted chat component and my header.

**Comment by jettb2** [comment]
I added the following code into GiftedChat and it fixed it for me
bottomOffset={33} 

<img width="1206" alt="Screen Shot 2022-09-03 at 6 08 35 PM" src="https://user-images.githubusercontent.com/92041686/188279108-9f762d97-aa6c-4f1a-a1d4-83455b89a3d9.png">

**Comment by KevinColemanInc** [comment]
The problem I found is wrapping GiftedChat (`v2.8.1`) in a "SafeAreaView" (or KeyboardAvoidingView, like chatGPT wanted) area pushes the input box too far above the keyboard. I think because internally, GiftChat has a "SaveView" and the wrapped "SafeAreaView" are compounding, resulting in double padding.

Unfortunately, when the keyboard is down and I removed the outer SafeAreaView, the input box is too close to the bottom part of the screen (putting the "up-swipe"/switch app slider on top of the input box).

What I ended up doing to fix this is wrap GiftedChat in SafeAreaView, but multiply `insets.bottom` by negative 1 (negating one layer of the SafeAreaView) when the keyboard is up.

```
import { useSafeAreaInsets } from 'react-native-safe-area-context'
  const insets = useSafeAreaInsets()

    <SafeAreaView style={{ flex: 1 }}>
      <GiftedChat
        ref={giftedChatRef}
        wrapInSafeArea={false}
        bottomOffset={Platform.OS === 'ios' ? insets.bottom * -1 : 0}
```

I do not know if the bottomOffset is correct for Android.
